use async jobs

coot opened the pull request:
The only sync command is `:PSCIDElist` as it outputs lots of input, and the omni-completion.
Otherwise all commands run in non-blocking way.

The syntastic is lacking support for async commands, `PSCIDErebuild` can be run in both blocking and non-blocking way.  The non-blocking way might be useful for people who don't want to use syntastic.

Commit:
use vim-jobs in s:callPsciIde

Commit:
callPscIde callbacks

Commit:
PSCIDErebuid

Commit:
PSCIDEcwd

Commit:
PSCIDEaddClause

Commit:
PSCIDEcaseSplit

Commit:
PSCIDEpursuit

Commit:
PSCIDElist

Commit:
s:callPsciIde callbacks

Commit:
s:callPsciIdeSync - synchronous call to psc-ide-client

Commit:
PSCIDElist and PSCIDEomni - use s:callPscIdeSync

Commit:
s:callPscIdeSync and callbacks

Commit:
s:PscIdeRetryCallback - sync version

Commit:
PSCIDEgoToDefinition

Commit:
PSCIDErebuid - use synchronus version in SyntasticMake

Commit:
PSCIDEend - async

Commit:
PSCIDErebuild - add callback argument

That's useful when one wants to setup error (async) handling
(syntastic only runs synchronously).

Commit:
error fix

Commit:
s:PSCIDEaddTypeAnnotationCallback - fix error

chexxor commented on ftplugin/purescript_pscide.vim:
```diff
@@ -142,8 +142,18 @@ function! PSCIDEend()
-  let input = {'command': 'quit'}
-  let resp = s:mysystem("psc-ide-client -p " . g:psc_ide_server_port, s:jsonEncode(input))
+  let filename = tempname()
+  call writefile([s:jsonEncode({'command': 'quite'})], filename)
```
Type? quite -> quit

chexxor commented on ftplugin/purescript_pscide.vim:
```diff
@@ -174,12 +184,19 @@ function! PSCIDEload(silent)
-    call s:log("PSCIDEload: Succesfully loaded modules: " . string(resp["result"]), loglevel)
+function! s:PSCIDEloadCallback(loglevel, resp)
+  if type(a:resp) == type({}) && a:resp['resultType'] ==# "success"
+    call s:log("PSCIDEload: Succesfully loaded modules: " . string(a:resp["result"]), a:loglevel)
```
Typo: Succesfully -> Successfully

chexxor commented on ftplugin/purescript_pscide.vim:
```diff
@@ -233,21 +248,29 @@ function! s:importIdentifier(id, module)
   "multiple possibilities
-  if type(resp) == type({}) && resp.resultType ==# "success" && type(resp.result[0]) == type({})
-    let choice = s:pickOption("Multiple possibilities to import " . ident, resp.result, "module")
+  call s:log("s:PSCIDEimportIndetifierCallback", 3)
```
Typo: PSCIDEimportIndetifierCallback -> PSCIDEimportIdentifierCallback

chexxor commented on ftplugin/purescript_pscide.vim:
```diff
@@ -346,28 +377,49 @@ function! s:goToDefinition(definedAt)
   let input = {'command': 'rebuild', 'params': {'file': filename}}
 
-  let resp = s:callPscIde(input, 0, 0)
+  if a:0 > 0 && type(a:1) == v:t_func
```
This condition is never reached, I think. I see this function is called in only one place with "0" argument.

coot replied:
Yes but that's useful if you want to define your own callback.  The syntastic plugin does not support async build processes, hence it blocks on writing the file.  I supply my custom function here.  I will make separate pull request to include it, after this patch gets in.

Expert from my `.vim/after/ftplugin/purescript.vim` file:
```viml
function! s:PSCIDEerrors(llist)
  let qflist = []
  for e in a:llist
    let eparts = split(e, ":")
    call add(
	  \   qflist
	  \ , { "bufnr": bufnr(eparts[1])
	  \   , "filename": eparts[1]
	  \   , "lnum": eparts[2]
	  \   , "col": eparts[3]
	  \   , "text": eparts[4]
	  \   , "type": eparts[0]
	  \   }
	  \ )
  endfor
  call setqflist(qflist)
endfunction
if g:psc_ide_syntastic_mode == 0
  com! PSCIDErebuild call PSCIDErebuild(1)
  augroup purescript
    au! BufWritePost *.purs call PSCIDErebuild(1, function("s:PSCIDEerrors"))
  augroup END
endif
```

coot replied:
By the way, thanks for the review :+1

chexxor reviewed (commented):
I'm curious if vim and neovim's async stuff has the same API - do you know? I'm curious which of these and which version of vim you're using.

I think it's possible to make a global flag which allows users to opt-in to async. Like `let g:psc_ide_async = 1` in my ~/.vimrc.

Each user-triggered function could check this flag and conditionally use async commands.

``` viml
command! PSCIDEimportIdentifier call PSCIDEimportIdentifier()
function! PSCIDEimportIdentifier()
  call s:importIdentifier(s:GetWordUnderCursor(), "")
 endfunction
function! s:importIdentifier(id, module)
  if g:psc_ide_async == 1
    call s:callPscIdeAsync( ..., {resp -> s:PSCIDEimportIdentifierCallback(ident, a:id, a:module, resp)} )
  else
    let resp = s:callPscIde(input, "Failed to import identifier " . ident, 0)
    call s:PSCIDEimportIdentifierCallback(ident, a:id, a:module, resp)
  endif
endfunction
```

We could change the name of `s:PSCIDEimportIdentifierCallback` to something more descriptive, like `s:...Handler` or `s:...Impl` or `s:...Cont` (continuation).

Commit:
include @chexxor review

Commit:
PSCIDEend - use exit_cb

Commit:
PSCIDEstart - start psc-ide-server as a job

Commit:
use viml json_encode and json_decode functions

coot commented:
Anybody to check it on windows and mac? Linux is covered :)

Commit:
fix - s:log requires two arguments

Commit:
s:PscIdeStartCallback - catch json decoding errors

* add err_cb in s:calPscIde
* fix how s:PscIdeStartCallback is called

Commit:
PSCIDEstart

* use lcd to start psc-ide-server in the right directory
* fix src paths

Commit:
clean code

FrigoEU commented:
Just wanted to let you know that this looks great! I'll need some time to check and test this and provide some feedback, but just wanted to let you know that I really appreciate the help. Thanks!

Commit:
Merge branch 'master' into jobs

Commit:
use viml filter function instead of s:findInListBy

Commit:
remove debug variable

Commit:
revert 8abd226

Commit:
finish merging upstream master

Commit:
use viml filter instead of s:take and s:drop

Commit:
s:getTypeCallback was masking error message when type was not found

coot commented:
Sure, take your time. That's a big  chunk of changes. I have only tested it on Linux.

Commit:
PSCIDEgoto

* use column
* use full path:
    previously length of current working directory was removed from the
    path - that does not work if the current working directory is not
    somewhere outside of a project

Commit:
README.md - add <buffer> and <C-U> to suggested nmap definitions

Commit:
buffer only commands

When using commnad-buffer the commands are only defined in the current
buffer - this way one does not polute the global name space with
filetype only commands.

Commit:
Don't use defined commands, instead call functions directly

Some users delete them and define their own set of commmands.

Commit:
filter requires using a copy

Commit:
filter results in PSCIDEimportIdentifierCallback & PSCIDEgoToDefinitionCallback

Commit:
PSCIDEgoToDefinitionCallback

Commit:
remove not neaded code

chexxor committed:
Merge remote-tracking branch 'coot/jobs' into master-mine

chexxor commented:
I've merged this branch with master to test it out. I'm having one problem: the `:PSCIDE*` commands aren't available in one .purs file (filetype is 'purescript'), but they *are* available in a different .purs file. @coot, have you seen this problem before?

It happens in gvim and vim, version `VIM - Vi IMproved 8.0 (2016 Sep 12, compiled...`. It seems to happen after I open .purs file, `:w`, see some error, `:PSCIDEload`, `:w`, see syntastic suggestions in gutter, then change to a different .purs file.

chexxor commented:
Actually, I think I fixed it by removing the `-buffer` flag from the `command!`s. I'll keep testing.

coot commented:
No, I haven't encountered that problem, but I delete all the commands and redefine them.  I always use `-buffer` (for commands) and `<buffer>` for commands defined in filetype plugins. They just don't pollute the global namespace.  That's important for tab completion for commands (and even more for maps, as in some configurations vim will wait until maps become unambiguous) - so your solution is not proper.

Check why the purescript filetype plugin is not executed.  What is the error that you see? Note that vim stops executing a script on first error - so if the commands are defined at the end vim will not execute that part of the script.

chexxor committed:
Fix getType command. Quote source file loaded.

chexxor commented:
@coot - The "getType" command wasn't working for me. See this patch for my fix, which is to get the "result" property from the result.
https://github.com/chexxor/psc-ide-vim/commit/43bbe0a4bba6b364e928a4aa67f2f3615fd619b4

Also, we need to quote the source-file glob we send to psc-ide-server, else it expands and is clipped when psc-ide-server receives it. It caused my "getType" command to fail for functions defined in bower_component libs, because I have so many libs there.

chexxor commented:
Also, the "gotoDefinition" function presumes the `definedAt` property in the result sets is non-null, but I see it is occasionally `v:null` (I don't know why).

kritzcreek commented:
`definedAt` can be null, it depends on whether the source files have been parsed yet and if the source positions have already been attached to the definitions parsed out of the externs files.

Also the source globs might not have caught the source file in question but it still could've appeared in the `output/` directory.

chexxor commented:
In my case, the module which contained the function to which I wanted to jump failed compilation.

coot commented:
@chexxor when I use quoted version of `"'src/**/*.purs'"` none of the src files is loaded. `PSCIDEload` sais:
`Loaded 39 modules and 0 source files`; then none of the commands works. However, when I leave it unquoted it correctly pics all the src files.  This does not affect `bower_components/**/*.purs` glob (quoted and unquoted version loads the same 39 modules).

coot commented:
I don't think vim expands globs when you use `job_start`, the [job-term](http://vim-jp.org/vimdoc-en/channel.html#job-term) option is not even implemented.

coot commented:
There is one more problem when opening multiple projects.  The server now, just fails to start and then `PSCIDEload` re-loads files from the existing server.  We should check if server fails to start and if it complains that it's already running we should use a different port.  Then we need to store the port in a buffer only variable.

chexxor commented:
@coot, yeah, I'll investigate further when I have time. Quoted and unquoted, either, never shows the same numbers of loaded modules and src files as when using master branch of psc-ide-vim.

kritzcreek commented:
Loaded Modules is about externs files, these get found with the output-path parameter and independent of source globs. You should only care about the "source files" number.

Commit:
do not quote globs for psc-ide-server

Commit:
PSCIDEgoToDefinitionCallback

* val.definedAt can be v:null
* simplify logic

coot commented:
@chexxor I found that the output of `:PSCIlist` is more reliable. It actually shows which modules are loaded. I pulled your changes (and unquoted the globs in 674d961 above).

Commit:
check type in PSCIDEgoToDefinitionCallback

FrigoEU commented:
Phew, I reviewed all the changes and it looks good! I tested it a little bit with vim8 and it seems to work very smoothly. The only problem is that I use neovim myself, and this causes two issues:

1. Lambda's are not supported in the current neovim release. They are merged in master, so that's good.
2. The job control API is different in neovim. There is a wrapper in the form of https://github.com/prabirshrestha/async.vim but I haven't tried it out yet. It looks solid though, I just haven't figured out yet how to write stuff to stdin into the job with this, but I suppose it shouldn't be a problem.

It's quite a pain that vim & neovim are diverging, but I guess it's a fact of life. I understand some (most?) users couldn't care less about neovim, but I actually really like it and am super excited about something like https://github.com/extr0py/oni which embeds neovim into something vscode-like. 

That said I'll look into using the async wrapper and seeing if I can get everything to work. I think I'm fine with requiring neovim master for lambda support. 

One las thing: Are many vim plugins already depending on vim8 jobs/lambdas? Just to know if requiring vim8 will annoy many users.

Thanks again for the monster work. Seeing as you rewrote most of psc-ide-vim, are you interested in becoming an official "contributor" so you get push access? Same goes for @chexxor .

coot commented:
I don't know of any stats how old version of vim are popular. vim-8 already has 445 patches (often when it grows over 1000 a new minor release is made, so we are in the middle way to 8.1). You can always point to the last release of the plugin that supports vim-7.4.

Unfortunately vim maintainer choose to implement async jobs in a different way.  I haven't used neovim much (only at the beginning, and it wasn't very stable back then).

Thanks, I will happily become official contributor, I will keep writting pull requests, that we should review, especially that you use neovim, and I might broke something using vim-8 only feature.

Commit:
PureScript 0.11

Commit:
Merge branch 'master' into jobs

Commit:
PSCIDElistImports

* show more identifiers and a qualifier
* use highlight colors

Commit:
s:ListImports - is used by getType

Commit:
s:PSCIDEimportIdentifierCallback

* fix: append `linesToAppend`
* fix: update line number in the value of windsaveview()

Commit:
s:formattype - use ∷ digraph

Commit:
s:PSCIDEloadCallback: do not access a:resp on error

It might not be a dictionary.

Commit:
s:PscIdeCallback - catch errors when decoding json

Commit:
Merge branch 'master' into jobs

coot commented:
@FrigoEU what about merging this changes, since it is a big PR it will be difficult to maintain it. Maybe you could branch it and make changes for neovim, or if it's problematic to maintain support for both at the same time, we may split it into two different repos.

Commit:
s:getType - include currentModule when filtering by module

Commit:
s:getType - include currentModule when filtering by module

Commit:
Merge branch 'jobs' of github.com:coot/psc-ide-vim into jobs

Commit:
fix: s:getType - put currentModule in params

FrigoEU committed:
Make async work on vim8 and neovim

FrigoEU commented:
Wow I actually pushed to your branch! Didn't know I could do that. Anyway, I spent the best part of a day getting this to work on both vim8 and neovim... Could you check it out and see what you think? I feel making all this stuff async has caused quite a bit of messiness, so I think it wouldn't be bad if we could clean up some stuff here and there, but we can always do that later. This PR has been bugging me for so long, so I would just like to get this merged finally!

coot commented:
Thanks I will check it.  I added lots of stuff in https://github.com/coot/psc-ide-vim
When we will be done with this PR, I will make another big PR...

coot commented:
For me it looks great.  You could actually publish this as a separate plugins on its own, since it may be useful for other projects making a bridge between vim and neovim.

I'd prefer to merge the branch first and  since I have another big PR (in preparation), but it should work in neovim without any trouble. Yes cleaning the code would also be nice, but lets do that after my PRs will get merged.

FrigoEU commented:
I was testing some more, and this doesn't seem to be working for me:

```
let key = fnamemodify(a:filename, ':s?'.dir.'/??') . "|" . string(a:lnr)
```

This changes 
/Users/Simon/psc-proj/rockschool/src/GUI/Component/Calendar.purs 
into
/UserSimon/psc-proj/rockschool/src/GUI/Component/Calendar.purs

Could you take a look at that?

FrigoEU committed:
Remove PSCIDEremoveImporQualifications, not relevant anymore

FrigoEU committed:
Fix boolean encoding in JSON

FrigoEU committed:
Fix async job stopping

coot commented:
Take a look at the [vim](https://github.com/coot/psc-ide-vim/tree/vim) branch at my clone repo.  I fixed that by just using `expand("%:p")`.  This got fixed in purescript-0.11.5.
I made additional changes there. I should document them somewhere (I list them here at the end).

Maybe we could just close this PR at take it from that branch - I merged your changes into that branch and it works fine.

Changes in the `vim` branch:
* support command arguments, for example `:PSCIDEtype [expr]` if `expr` is given type of `expr` is returned otherwise current cursor position is used
* `PSCIDEapplySuggestions!` - support bang: apply all suggestions
* `PSCIDEload!` - support bang: reset the server and then load modules
* `PSCIDEsearch` command: search for expr
* user completion `i_^X^U` (check `:h 'completefunc'`) - another completion function, this finds as much completions as possible using the flex matcher - useful when you'll looking for a command
* I added some options for omnicompletion
  -  `g:psc_ide_omnicompletion_filter_modules`
  -  `g:psc_ide_omnicompletion_sort_by` (`flex` /  `identifier` / `module`)
  -  `g:psc_ide_omnicompletion_prefix_filter` (if set to false it fixes `flex` matcher, which 
previously did not work)
  -  I added support for grouping modules (purescript-0.11.5 new feature)
* I extended support for `:PSCIDEimportIdentifier`, there are some options:
  - `g:psc_ide_filter_prelude_modules`(if true only shows results from `Prelude` rather than all the results from modules that are part of it)
  - `g:psc_ide_filter_submodules` - this emulates grouping of omnicompletion which is not yet in implemented for `import` command - I actually have to check it, since I did it before 0.11.5 and maybe it works.
  - import identifier now is just replacing the file, there were various edge cases that I didn't want to fix with the previous algorithm.  I make it the way that the cursor position is preserved.
* `:PSCIDEsplitCase` now expects an argument, rather than ask for it during execution.
* I fixed support for buffer local commands
* I added auto start mechanism (starting server, loading files into it, maybe we should also add initial rebuild so quickfix list is filled)
* I added titles to quickfix / location list used (`PureScript Errors`, `PureScript Search` - in `:PSCIDEsearch` command)
* I restructured the script so that it avoids reevaluation (one can get errors if the script evaluates during execution of one of its functions, since vim does not allow to replace a function when it is executed), the proper solution for that is to use autoload directory and move all the functions there, but I didn't want to restructure the code too much before `jobs` get merged
* logging: log json encoded commands, so you just need to copy paste them into terminal (at least on linux when you use string function, you don't get a valid json)
* `:PSCIDEerr` command which shows the current error (multiline) - previously the plugin was removing newlines, but I get rid of that, since vim & neovim already handles that)
* `:PSCIDEtype` adds current module to imported modules, so you can query results about expressions from the current module

coot commented:
We should move the documentation from `README.md` into vim doc format...

FrigoEU commented:
Wowzers, looks insane. Sure if you merged all the small things I fixed today into your master, just make a PR for that and I'll merge that!

coot commented:
ok
